### PR TITLE
LA::d::Vector import

### PR DESCRIPTION
**Pull request by Rombur**
This replaces #8389 With this PR, we can directly import a `LA::d::Vector` from one `MemorySpace` to another.

### CODE REVIEW

File: doc/news/changes/minor/20190724BrunoTurcksin [hunk 1]
```diff
@@ -0,0 +1,5 @@
+Minor:: New function LinearAlgebra::distributed::Vector::import() to directly
```
Comment (kronbichler) [inline doc/news/changes/minor/20190724BrunoTurcksin]:
```suggestion
New: The function LinearAlgebra::distributed::Vector::import() allows to
```

File: include/deal.II/lac/la_parallel_vector.h [hunk 2]
```diff
@@ -614,6 +614,20 @@ namespace LinearAlgebra
       void
       copy_locally_owned_data_from(const Vector<Number2, MemorySpace> &src);
 
+      /**
+       * Import all the elements present in the distributed vector @p src.
+       * VectorOperation::values @p operation is used to decide if the elements
+       * in @p V should be added to the current vector or replace the current
+       * elements.
+       *
+       * @note The partitioner of the two distributed vectors need to be the
```
Comment (kronbichler) [inline include/deal.II/lac/la_parallel_vector.h]:
```suggestion
       * @note The partitioners of the two distributed vectors need to be the
```

File: include/deal.II/lac/la_parallel_vector.h [hunk 3]
```diff
@@ -614,6 +614,20 @@ namespace LinearAlgebra
       void
       copy_locally_owned_data_from(const Vector<Number2, MemorySpace> &src);
 
+      /**
+       * Import all the elements present in the distributed vector @p src.
+       * VectorOperation::values @p operation is used to decide if the elements
+       * in @p V should be added to the current vector or replace the current
+       * elements.
```
Comment (kronbichler) [inline include/deal.II/lac/la_parallel_vector.h]: Maybe you can explicitly add something like "The main purpose of this function is to get data from one memory space, e.g. CUDA, to the other, e.g. the host." It is clear from the template arguments and implicitly by the description above, but I think that being even more clear on this topic is never wrong.

**Comment by Rombur** [comment]
Done

File: include/deal.II/lac/la_parallel_vector.templates.h [hunk 4]
```diff
@@ -727,6 +727,26 @@ namespace LinearAlgebra
...
+    template <typename MemorySpaceType2>
+    void
+    Vector<Number, MemorySpaceType>::import(
+      const Vector<Number, MemorySpaceType2> &src,
+      VectorOperation::values                 operation)
+    {
+      Assert(src.partitioner.get() != nullptr, ExcNotInitialized());
+      Assert(partitioner->locally_owned_range() ==
+               src.partitioner->locally_owned_range(),
+             ExcMessage("Locally unknown indices should be identical."));
```
Comment (masterleinad) [inline include/deal.II/lac/la_parallel_vector.templates.h]: Did you mean "Locally owned ..."?

**Comment by masterleinad** [comment]
```
419/592 Test #1268: all-headers/lac/vector_operations_internal.h.debug ...................***Failed    5.80 sec
In file included from /jenkins/workspace/dealii_PR-8412/tests/all-headers/test_header.cc:17:0:
/jenkins/workspace/dealii_PR-8412/include/deal.II/lac/vector_operations_internal.h:1667:9: error: ‘VectorOperation’ has not been declared
         VectorOperation::values /*operation*/,
         ^
/jenkins/workspace/dealii_PR-8412/include/deal.II/lac/vector_operations_internal.h:2023:14: error: 'VectorOperation’ has not been declared
              VectorOperation::values operation,
              ^
/jenkins/workspace/dealii_PR-8412/include/deal.II/lac/vector_operations_internal.h:2023:38: error: expected ‘,’ or ‘...’ before ‘operation’
              VectorOperation::values operation,
                                      ^
/jenkins/workspace/dealii_PR-8412/include/deal.II/lac/vector_operations_internal.h: In static member function ‘static void dealii::internal::VectorOperations::functions<Number, Number2, dealii::MemorySpace::Host>::import(const std::shared_ptr<dealii::parallel::internal::TBBPartitioner>&, dealii::internal::VectorOperations::size_type, int)’:
/jenkins/workspace/dealii_PR-8412/include/deal.II/lac/vector_operations_internal.h:2033:13: error: ‘operation’ was not declared in this scope
         if (operation == VectorOperation::insert)
             ^
/jenkins/workspace/dealii_PR-8412/include/deal.II/lac/vector_operations_internal.h:2033:26: error: ‘VectorOperation’ has not been declared
         if (operation == VectorOperation::insert)
                          ^
/jenkins/workspace/dealii_PR-8412/include/deal.II/lac/vector_operations_internal.h:2035:49: error: ‘v_data’ was not declared in this scope
             copy(thread_loop_partitioner, size, v_data, data);
                                                 ^
/jenkins/workspace/dealii_PR-8412/include/deal.II/lac/vector_operations_internal.h:2035:57: error: ‘data’ was not declared in this scope
             copy(thread_loop_partitioner, size, v_data, data);
                                                         ^
/jenkins/workspace/dealii_PR-8412/include/deal.II/lac/vector_operations_internal.h:2037:31: error: ‘VectorOperation’ has not been declared
         else if (operation == VectorOperation::add)
                               ^
/jenkins/workspace/dealii_PR-8412/include/deal.II/lac/vector_operations_internal.h:2039:55: error: ‘v_data’ was not declared in this scope
             add_vector(thread_loop_partitioner, size, v_data, data);
                                                       ^
/jenkins/workspace/dealii_PR-8412/include/deal.II/lac/vector_operations_internal.h:2039:63: error: ‘data’ was not declared in this scope
             add_vector(thread_loop_partitioner, size, v_data, data);
```

**Pull request merged**